src/razor/src/completion/completionHandler.ts

```diff
@@ -0,0 +1,185 @@
+                delegatedCompletionParams.identifier.textDocumentIdentifier.uri,
+                true
+            );
+            const razorDocument = await this.documentManager.getDocument(razorDocumentUri);
+            if (razorDocument === undefined) {
+                return CompletionHandler.emptyCompletionList;
+            }
+
+            const textDocument = await vscode.workspace.openTextDocument(razorDocumentUri);
+
+            let virtualDocument: any;
+            if (delegatedCompletionParams.projectedKind === LanguageKind.Html) {
+                virtualDocument = razorDocument.htmlDocument;
+            } else if (delegatedCompletionParams.projectedKind === LanguageKind.CSharp) {
+                virtualDocument = razorDocument.csharpDocument;
+            } else {
+                // TODO: equivalent of Debug.Fail?
+                return CompletionHandler.emptyCompletionList;
+            }
+
+            // TODO: Should we check for null or undefined virtual document like we do in C# code?
+
+            const synchronized = await this.documentSynchronizer.trySynchronizeProjectedDocument(
+                textDocument,
+                virtualDocument,
+                delegatedCompletionParams.identifier.version,
+                token
+            );
+            if (!synchronized) {
+                return CompletionHandler.emptyCompletionList;
+            }
+
+            const virtualDocumentUri = UriConverter.serialize(virtualDocument.uri);
+
+            delegatedCompletionParams.identifier.textDocumentIdentifier.uri = virtualDocumentUri;
+
+            // TODO: support for provisional edits
+
+            // "@" is not a valid trigger character for C# / HTML and therefore we need to translate
+            // it into a non-trigger invocation.
+            const modifiedTriggerCharacter =
+                delegatedCompletionParams.context.triggerCharacter === '@'
+                    ? undefined
+                    : delegatedCompletionParams.context.triggerCharacter;
+            const triggerKind =
+                delegatedCompletionParams.context.triggerCharacter === '@'
+                    ? CompletionTriggerKind.Invoked
+                    : delegatedCompletionParams.context.triggerKind;
+
+            let completions: vscode.CompletionList | vscode.CompletionItem[];
+
+            if (delegatedCompletionParams.projectedKind === LanguageKind.CSharp) {
+                const params: CompletionParams = {
+                    context: {
+                        triggerKind: triggerKind,
+                        triggerCharacter: modifiedTriggerCharacter,
+                    },
+                    textDocument: {
+                        uri: virtualDocumentUri,
+                    },
+                    position: delegatedCompletionParams.projectedPosition,
+                };
+
+                completions = await vscode.commands.executeCommand<vscode.CompletionList | vscode.CompletionItem[]>(
```

Comment:
The return type here looks wrong. These are the VS Code types, and I think they should be the LSP types. It might not matter right now in this instance, but I've definitely seen VS Code types and LSP types differ, so we can lose data.

Reply:
This was copied verbatim from https://github.com/dotnet/vscode-csharp/blob/fbdd604dda7a686e05bf5cba25c7a7b2670233de/src/razor/src/completion/razorCompletionItemProvider.ts#L42

Was that a bug in the old completion code? I think you might be right at least for C#. I think C# returns the LSP types, but HTML command returns VSCode types. If I passed on C# returned data back to Razor server without conversion, all was well. If I passed HTML retrurned data for Razor server without conversion, I got JSON deserialization exception. That was fixed when I added conversion from VScode to LSP types. 

So I'm guessing Roslyn is returning LSP types but VSCode HTML command is returning vscode types.

Reply:
> Was that a bug in the old completion code

Yes, I think so. There might not have been any real issues it caused if the serialization happened to be correct, but it wouldn't be something we'd want to bank on.

Reply:
Sounds good, changed that for Roslyn completion. I think using vscode types for HTML is the right thing to do though, we just have to convert them to LSP types ourselves before giving them back to the server. Let me know if you disagree. See next commit.